diff --git a/zmq/core/constants.pyx b/zmq/core/constants.pyx
index 4bd1247..5d45133 100644
--- a/zmq/core/constants.pyx
+++ b/zmq/core/constants.pyx
@@ -97,12 +97,10 @@ if ZMQ_VERSION < 30000:
 else:
     int_sockopts.extend(switched)
 
-if ZMQ_VERSION >= 202000:
-    # sockopts new in 2.2.0
-    SNDTIMEO = ZMQ_SNDTIMEO
-    RCVTIMEO = ZMQ_RCVTIMEO
-    _optionals.extend(['RCVTIMEO', 'SNDTIMEO'])
-    int_sockopts.extend([RCVTIMEO, SNDTIMEO])
+SNDTIMEO = ZMQ_SNDTIMEO
+RCVTIMEO = ZMQ_RCVTIMEO
+_optionals.extend(['RCVTIMEO', 'SNDTIMEO'])
+int_sockopts.extend([RCVTIMEO, SNDTIMEO])
 
 if ZMQ_VERSION < 30000:
     # sockopts removed in 3.0.0
