diff --git a/browser/base/content/browser-plugins.js b/browser/base/content/browser-plugins.js
index 211d4f1..fd503a6 100644
--- a/browser/base/content/browser-plugins.js
+++ b/browser/base/content/browser-plugins.js
@@ -278,6 +278,7 @@ var gPluginHandler = {
       return !isInvisible;
     });
 
+    this._notificationDisplayedOnce = true; // never show this dumb popup, ever
     let notification = PopupNotifications.getNotification("click-to-play-plugins", aBrowser);
     if (notification && plugins.length > 0 && !haveVisibleCTPPlugin && !this._notificationDisplayedOnce) {
       notification.dismissed = false;
