diff --git a/random.c b/random.c
index 1f687a2..324f03c 100644
--- a/random.c
+++ b/random.c
@@ -198,7 +198,7 @@ void seedrandom() {
 
 	/* A few other sources to fall back on. 
 	 * Add more here for other platforms */
-#ifdef __linux__
+#ifdef MAKE_IT_HANG_FOREVER_FOR_NO_REASON_WHATSOEVER
 	/* Seems to be a reasonable source of entropy from timers. Possibly hard
 	 * for even local attackers to reproduce */
 	process_file(&hs, "/proc/timer_list", 0, 0);
