diff --git a/src/cryptsetup.c b/src/cryptsetup.c
index 73f4c24..3063878 100644
--- a/src/cryptsetup.c
+++ b/src/cryptsetup.c
@@ -64,7 +64,7 @@ static int opt_random = 0;
 static int opt_urandom = 0;
 static int opt_dump_master_key = 0;
 static int opt_shared = 0;
-static int opt_allow_discards = 0;
+static int opt_allow_discards = 1;
 
 static const char **action_argv;
 static int action_argc;
@@ -1281,11 +1281,8 @@ int main(int argc, const char **argv)
 	if (opt_allow_discards &&
 	    strcmp(aname, "luksOpen") &&
 	    strcmp(aname, "create") &&
-	    strcmp(aname, "loopaesOpen")) {
-		usage(popt_context, EXIT_FAILURE,
-		      _("Option --allow-discards is allowed only for luksOpen, loopaesOpen and create operation.\n"),
-		      poptGetInvocationName(popt_context));
-	}
+	    strcmp(aname, "loopaesOpen"))
+		opt_allow_discards = 0;
 
 	if (opt_key_size &&
 	   strcmp(aname, "luksFormat") &&
