diff --git a/lib/refcounted.h b/lib/refcounted.h
index 9eb638e..52493e2 100644
--- a/lib/refcounted.h
+++ b/lib/refcounted.h
@@ -25,6 +25,8 @@
  USA
 ***********************************************************************/
 
+#include <stddef.h>
+
 #if !defined(MYSQLPP_REFCOUNTED_H)
 #define MYSQLPP_REFCOUNTED_H
 
@@ -89,7 +91,7 @@ public:
 	refs_(0)
 	{
 	}
-	
+
 	/// \brief Standard constructor
 	///
 	/// \param c A pointer to the object to be managed.  If you pass 0,
@@ -184,13 +186,13 @@ public:
 	T* operator ->() const
 	{
 		return counted_;
-	}	
+	}
 
 	/// \brief Dereference the smart pointer
 	T& operator *() const
 	{
 		return *counted_;
-	}	
+	}
 
 	/// \brief Returns the internal raw pointer converted to void*
 	///
@@ -244,7 +246,7 @@ public:
 	{
 		std::swap(counted_, other.counted_);
 		std::swap(refs_, other.refs_);
-	}	
+	}
 
 private:
 	/// \brief Pointer to the reference-counted object
@@ -262,4 +264,3 @@ private:
 } // end namespace mysqlpp
 
 #endif // !defined(MYSQLPP_REFCOUNTED_H)
-
diff --git a/ssx/genv2.cpp b/ssx/genv2.cpp
index b6c75a1..4cb2a34 100644
--- a/ssx/genv2.cpp
+++ b/ssx/genv2.cpp
@@ -35,6 +35,7 @@
 #include <iostream>
 #include <fstream>
 #include <typeinfo>
+#include <cstring>
 
 using namespace std;
 
