diff --git a/server/server.py b/server/server.py
index 773fdf0..3f164d1 100644
--- a/server/server.py
+++ b/server/server.py
@@ -454,24 +454,6 @@ class Server:
 		else:
 			os._exit(0)		 # Exit parent of the first child.
 		
-		# Close all open files.  Try the system configuration variable, SC_OPEN_MAX,
-		# for the maximum number of open files to close.  If it doesn't exist, use
-		# the default value (configurable).
-		try:
-			maxfd = os.sysconf("SC_OPEN_MAX")
-		except (AttributeError, ValueError):
-			maxfd = 256	   # default maximum
-	
-		for fd in range(0, maxfd):
-			try:
-				os.close(fd)
-			except OSError:   # ERROR (ignore)
-				pass
-	
-		# Redirect the standard file descriptors to /dev/null.
-		os.open("/dev/null", os.O_RDONLY)	# standard input (0)
-		os.open("/dev/null", os.O_RDWR)		# standard output (1)
-		os.open("/dev/null", os.O_RDWR)		# standard error (2)
 		return True
 
 
